zerver/lib/url_preview/preview.py

```diff
@@ -0,0 +1,63 @@
+import re
+import logging
+import traceback
+from six import text_type
+from typing import Any, Optional
+from typing.re import Match
+import requests
+from django.conf import settings
+from zerver.lib.cache import cache_with_key, get_cache_with_key
+from zerver.lib.bugdown import testing_mocks
+from .oembed import get_oembed_data
+from .parsers import OpenGraphParser, GenericParser
+
+
+CACHE_NAME = "database"
+link_regex = re.compile(
+    r'^(?:http)s?://'  # http:// or https://
+    r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
+    r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
+    r'(?::\d+)?'  # optional port
+    r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+
+
+def is_link(url):
+    # type: (text_type) -> Match[text_type]
+    return link_regex.match(str(url))
+
+
+def cache_key_func(url):
+    # type: (text_type) -> text_type
+    return url
+
+
+@cache_with_key(cache_key_func, cache_name=CACHE_NAME, with_statsd_key="urlpreview_data")
+def get_link_embed_data(url, maxwidth=640, maxheight=480):
+    # type: (text_type, Optional[int], Optional[int]) -> Any
+    if not is_link(url):
+        return None
+    if settings.TEST_SUITE:
+        return testing_mocks.fake_urlembed_data()
+    try:
+        data = get_oembed_data(url, maxwidth=maxwidth, maxheight=maxheight)
+    except requests.exceptions.RequestException:
+        msg = 'Unable to fetch information from url {0}, traceback: {1}'
+        logging.error(msg.format(url, traceback.format_exc()))
+        return None
+    if data:
+        return data
+    response = requests.get(url)
```

Comment:
Can you add a comment explaining the logic here?  Is it basically that we first try to use python-oembed to get the oembed data, but if that isn't available, we check open graph and then just the page title data?